### PR TITLE
gh-103204: `http.server` - Enforce that HTTP version numbers must consist only of digits

### LINKED ISSUE #103204: `http.server` parses HTTP version numbers too permissively.
# `http.server` parses HTTP version numbers too permissively.
`http.server` accepts request lines with HTTP version numbers that have `'_'`, `'+'`,  and `'-'`.

# Reproduction steps:
(Requires netcat)
```bash
python3 -m http.server --bind 127.0.0.1
printf 'GET / HTTP/-9_9_9.+9_9_9\r\n\r\n' | nc 127.0.0.1 8000
```

# Justification
Here are the `HTTP-version` definitions from each of the three HTTP RFCs:
- RFC 2616:
```
HTTP-Version   = "HTTP" "/" 1*DIGIT "." 1*DIGIT
```
- RFC 7230:
```
HTTP-version  = HTTP-name "/" DIGIT "." DIGIT
HTTP-name     = %x48.54.54.50 ; "HTTP", case-sensitive
```
- RFC 9112:
```
HTTP-version  = HTTP-name "/" DIGIT "." DIGIT
HTTP-name     = %s"HTTP"
```
I understand allowing multiple digits for backwards-compatibility with RFC 2616, but I don't think it makes sense to let the specifics of `int` leak out into the world. We should at least ensure that only digits are permitted in HTTP version numbers.

# My environment
- CPython 3.12.0a6+
- Operating system and architecture: Arch Linux on x86_64

### Linked PRs
* gh-103205
* gh-104438

### PR DESCRIPTION
**Pull request by kenballus**
# gh-103204: Enforce that HTTP version numbers must consist only of digits

Currently, `http.server` accepts requests with version numbers preceded with `'+'` or `'-'`, as well as those with `'_'` between digits. I am not aware of any other HTTP server that accepts such requests. This PR makes it reject those requests.

Fixes #103204.

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Could you add a unit test?

File: Lib/http/server.py [hunk 1]
```diff
@@ -300,6 +300,9 @@ def parse_request(self):
                 #   - Leading zeros MUST be ignored by recipients.
                 if len(version_number) != 2:
                     raise ValueError
+                for component in version_number:
+                    if not component.isdigit():
+                        raise ValueError
```
Comment (arhadthedev) [inline Lib/http/server.py]: Will this work? I have no Python machine on hands to check myself.

```suggestion
                if any(not component.isdigit() for component in version_number):
                    raise ValueError
```

Comment (kenballus) [reply Lib/http/server.py]: lgtm

**Comment by arhadthedev** [comment]
@gpshead (as a more-than-twice committer into `Lib/http/server.py`)

**Comment by gpshead** [comment]
From a protocol standpoint this PR is pedantically correct, we shouldn't have accepted other values. Not that anyone is ever going to send them. I've also added to this PR a length constraint on the values accepted for good measure.

**Comment by JelleZijlstra** [comment]
`.isdigit()` allows any Unicode digit, so I was worried we could get into some shenanigans with weird non-ASCII digits. But it seems that the way we get this version string, that's not possible.

**Pull request merged**